onnx/shape_inference/shape_inference.h

```diff
@@ -0,0 +1,92 @@
+#pragma once
+
+#include "onnx/proto_utils.h"
+#include "onnx/defs/schema.h"
+
+namespace ONNX_NAMESPACE { namespace shape_inference {
+
+struct InferenceContextImpl : public InferenceContext {
+  virtual const AttributeProto* getAttribute(std::string name) {
+    auto iter = attributesByName_.find(name);
+    if (iter == attributesByName_.end()) {
+      return nullptr;
+    } else {
+      return iter->second;
+    }
+  }
+  virtual const std::vector<const TypeProto_Tensor *>& getAllInputTypes() {
+    return allInputTypes_;
+  }
+  virtual void setAllOutputTypes(std::vector<TypeProto_Tensor>&& allOutputTypes) {
+    allOutputTypes_ = allOutputTypes;
+  }
+  std::unordered_map<std::string, const AttributeProto *> attributesByName_;
+  std::vector<const TypeProto_Tensor*> allInputTypes_;
+  std::vector<TypeProto_Tensor> allOutputTypes_;
+};
+
+void InferShapes(ONNX_NAMESPACE::ModelProto& m) {
+  auto g = m.mutable_graph();
+
+  std::unordered_map<std::string, TypeProto_Tensor> valueTypesByName;
+  for (auto vi : g->value_info()) {
+    valueTypesByName[vi.name()] = vi.type().tensor_type();
+  }
+  for (auto vi : g->input()) {
+    valueTypesByName[vi.name()] = vi.type().tensor_type();
+  }
+  for (auto vi : g->output()) {
+    valueTypesByName[vi.name()] = vi.type().tensor_type();
+  }
+
+  for (auto n : g->node()) {
+    auto schema = OpSchemaRegistry::Schema(n.op_type());
+    if (schema) {
+      InferenceContextImpl ctx;
+
+      {
+        std::unordered_map<std::string, const AttributeProto *> attributesByName;
+        for (auto& attr : n.attribute()) {
+          attributesByName[attr.name()] = &attr;
+        }
+        ctx.attributesByName_ = std::move(attributesByName);
+      }
+
+      {
+        std::vector<const TypeProto_Tensor *> allInputTypes;
+
+        for (auto input : n.input()) {
+          auto iter = valueTypesByName.find(input);
+          if (iter != valueTypesByName.end()) {
+            allInputTypes.push_back(&iter->second);
+          } else {
+            allInputTypes.push_back(nullptr);
+          }
+        }
+
+        ctx.allInputTypes_ = std::move(allInputTypes);
+      }
+
+      auto shi = schema->GetShapeInferenceFunction();
+      if (shi) {
+        shi(ctx);
+      }
+
+      auto& allOutputTypes = ctx.allOutputTypes_;
+      if (allOutputTypes.size() != n.output_size()) {
+        continue;
+      }
+      for (int i = 0; i < n.output_size(); i++) {
+        auto output = n.output(i);
+        if (valueTypesByName.find(output) == valueTypesByName.end()) {
+          valueTypesByName[output] = allOutputTypes[i];
```

Comment:
some follow ups are better to just include in code as TODO. E.g. in this place - regarding merging of partially specified shapes

Reply:
++ I'll add it here as well